src/sobol_sensitivity.jl

```diff
@@ -0,0 +1,144 @@
+        p = [0.0 for i in p_range]
+        j = 1
+        for i in 1:length(p_range)
+            if i in indices
+                p[i] = p_fixed[j]
+                j += 1
+            else
+                p[i] = (p_range[i][2] -p_range[i][1])*rand() + p_range[i][1]
+            end
+        end
+    end
+    p
+end
+
+function calc_mean_var(f,p_range,N)
+    y1 = give_rand_y(f,p_range)
+    y0 = [0.0 for j in 1:length(y1)]
+    v = [0.0 for j in 1:length(y1)]
+    if length(size(y1)) != 1
+        y0 = reshape(y0,size(y1)[1],size(y1)[2])
+        v = reshape(v,size(y1)[1],size(y1)[2])
+    end
+    for i in 1:N
+        y1 = give_rand_y(f,p_range)
+        y0 .+= y1
+        y1 = give_rand_y(f,p_range)
+        v .+= y1.^2
+    end
+    y0 = y0./N
+    y0_sq = [i.^2 for i in y0]
+    v = v./N .- y0_sq
+    y0,v
+end
+
+function first_order_var(f,p_range,N,y0)
+    ys = []
+    for i in 1:length(p_range)
+        y = [0.0 for i in 1:length(y0)]
+        if length(size(y0)) != 1
+            y = reshape(y,size(y0)[1],size(y0)[2])
+        end
+        for j in 1:N
+            p2 = give_rand_p(f,p_range)
+            p1 = give_rand_p(f,p_range,[p2[i]],[i])
+            yer = Array(f(p1)) .* Array(f(p2))
+            y .+= yer
+        end
+        y = y./N - y0.^2
+        push!(ys,copy(y))
+    end
+    ys
+end
+
+function second_order_var(f,p_range,N,y0)
+    ys = []
+    for i in 1:length(p_range)
+        for j in i+1:length(p_range)
+            y = [0.0 for i in 1:length(y0)]
+            if length(size(y0)) != 1
+                y = reshape(y,size(y0)[1],size(y0)[2])
+            end
+            for k in 1:N
+                p2 = give_rand_p(f,p_range)
+                p1 = give_rand_p(f,p_range,[p2[i],p2[j]],[i,j])
+                y .+=  Array(f(p1)) .* Array(f(p2)) 
+            end
+            y = y./N - y0.^2
+            push!(ys,copy(y))
+        end
+    end
+    ys_frst_order = first_order_var(f,p_range,N,y0)
+    j = 1
+    for i in 1:length(p_range)
+        for k in i+1:length(p_range)
+            ys[j] = ys[j] .- ( ys_frst_order[i] .+ ys_frst_order[k] )
+            j += 1
+        end
+    end
+    ys
+end
+
+
+function total_var(f,p_range,N,y0)
+    ys = []
+    for i in 1:length(p_range)
+        y = [0.0 for i in 1:length(y0)]
+        if length(size(y0)) != 1
+            y = reshape(y,size(y0)[1],size(y0)[2])
+        end
+        for j in 1:N
+            p_fixed_all = []
+            p_fixed_indices = []
+            p2 = give_rand_p(f,p_range)
+            for j in 1:length(p2)
+                if j != i
+                    push!(p_fixed_all,p2[j])
+                    push!(p_fixed_indices,j)
+                end
+            end
+            p1 = give_rand_p(f,p_range,p_fixed_all,p_fixed_indices) 
+            yer = Array(f(p1)) .* Array(f(p2))
+            y .+= yer
+        end
+        y = y./N - y0.^2
+        push!(ys,copy(y))
+    end
+    ys
+end
+
+function sobol_sensitivity(f,p_range,N,order=0)
```

Comment:
default to second order.

Reply:
no. The total effects are far too expensive in most cases so they aren't really ever used. Sobol is useful over things like Morris since it can get the second order cases, so it's usually used because of its ability to get the second order interactions only. The total variance estimate then gives you a measure for how much the higher order interactions matter, which is usually low and justifies only doing the second order.